WordPress/src/test/java/org/wordpress/android/ui/prefs/categories/detail/CategoryDetailViewModelTest.kt

```diff
@@ -1,25 +1,157 @@
+
     private val dispatcher: Dispatcher = mock()
+    private val networkUtilsWrapper: NetworkUtilsWrapper = mock()
+    private val getCategoriesUseCase: GetCategoriesUseCase = mock()
+    private val addCategoryUseCase: AddCategoryUseCase = mock()
+    private val resourceProvider: ResourceProvider = mock()
+    private val selectedSiteRepository: SelectedSiteRepository = mock()
 
-    private lateinit var viewModel: CategoryDetailViewModel
+    private val uiStates = mutableListOf<UiState>()
+    private val onCategoryPushStates = mutableListOf<Event<CategoryUpdateUiState>>()
+
+    private val siteModel: SiteModel = mock()
+    private val topLevelCategory = "Top level"
 
     @Before
     fun setUp() {
-        viewModel = CategoryDetailViewModel(TEST_DISPATCHER, dispatcher)
+        setUpMockResponse()
+        viewModel = CategoryDetailViewModel(
+                TEST_DISPATCHER,
+                networkUtilsWrapper,
+                getCategoriesUseCase,
+                addCategoryUseCase,
+                resourceProvider,
+                dispatcher,
+                selectedSiteRepository
+        )
+        setUpUiStateObservers()
+    }
+
+    private fun setUpUiStateObservers() {
+        viewModel.uiState.observeForever { if (it != null) uiStates += it }
+        viewModel.onCategoryPush.observeForever { if (it != null) onCategoryPushStates += it }
+    }
+
+    private fun setUpMockResponse() {
+        whenever(selectedSiteRepository.getSelectedSite()).thenReturn(siteModel)
+        whenever(resourceProvider.getString(R.string.top_level_category_name)).thenReturn(topLevelCategory)
+        whenever(getCategoriesUseCase.getSiteCategories(siteModel)).thenReturn(arrayListOf())
+    }
+
+    @Test
+    fun `when vm starts, then default parent category is selected`() {
+        viewModel.start()
+        assertEquals(topLevelCategory, uiStates.first().categories[0].name)
+    }
+
+    @Test
+    fun `when vm starts, then submit button is shown and disabled`() {
+        viewModel.start()
+        assertFalse(uiStates.first().submitButtonUiState.enabled)
+        assertTrue(uiStates.first().submitButtonUiState.visibility)
+    }
+
+    @Test
+    fun `given category name is updated, when vm starts, then submit button is enabled`() {
+        viewModel.start()
+        viewModel.onCategoryNameUpdated("category name")
+        assertTrue(uiStates.last().submitButtonUiState.enabled)
     }
 
     @Test
-    fun `sample test`() {
-    } // TODO:
+    fun ` given parent category is updated, when vm starts, then ui state is updated`() {
+        val categoryNode = CategoryNode(1, 1, "Category")
+        whenever(getCategoriesUseCase.getSiteCategories(siteModel)).thenReturn(arrayListOf(categoryNode))
+        viewModel.start()
+        val selectedCategoryParent = 2
+        viewModel.onParentCategorySelected(selectedCategoryParent)
+        assertEquals(selectedCategoryParent, uiStates.last().selectedParentCategoryPosition)
```

Comment:
🔍 `given/ when/ then` sections can be split by a blank line for ease of readability (applies to all test cases in the class).

Reply:
@ashiagr I am sorry, but I am not sure that splitting the `given/when/then` by blank line instead of `,` will improve the readability. And also, correct me if I am wrong. Since most of the tests have `given/ when/ then` split by `,`, wouldn't it be better if we kept it that way?

Reply:
I was referring to sections (see highlighted lines) and not the test description. 😅 Test description clauses are perfectly split by commas. 

Notice blank lines between `given, when, and then` sections or `arrange, act, and assert` sections. Hope it clarifies. 🤞 

```
@Test
fun `given parent category is updated, when vm starts, then ui state is updated`() {
    val categoryNode = CategoryNode(1, 1, "Category")
    whenever(getCategoriesUseCase.getSiteCategories(siteModel)).thenReturn(arrayListOf(categoryNode))

    viewModel.start()
        
    val selectedCategoryParent = 2
    viewModel.onParentCategorySelected(selectedCategoryParent)
    assertEquals(selectedCategoryParent, uiStates.last().selectedParentCategoryPosition)
}
```

Reply:
@ashiagr 😅 I didn't understand it first, so sorry. I think I got it now, but just to be clear before I push the changes, 

```
  @Test
    fun ` given parent category is updated, when vm starts, then ui state is updated`() {
        //arrange
        val categoryNode = CategoryNode(1, 1, "Category")
        whenever(getCategoriesUseCase.getSiteCategories(siteModel)).thenReturn(arrayListOf(categoryNode))

        //act
        viewModel.start()
        val selectedCategoryParent = 2
        viewModel.onParentCategorySelected(selectedCategoryParent)

        //assert
        assertEquals(selectedCategoryParent, uiStates.last().selectedParentCategoryPosition)
    }
```

Does this looks correct?. Since         `viewModel.onParentCategorySelected(selectedCategoryParent)` is an act shouldn't it be in the same section as  `viewModel.start()`?.

Reply:
Perfect understanding this time! 🙌 

Just that you can remove the comments:
//arrange
//act
//assert

as now the sections are separated by a single blank line.

Reply:
@ashiagr Thanks for taking the time and letting me know this. This really improves the code readability. I have updated the sections by a single blank line in this commit 33117a4.